compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -931,7 +931,7 @@ impl<'p, 'tcx> PatternColumn<'p, 'tcx> {
             let specialized = pat.specialize(pcx, &ctor);
             for (subpat, column) in specialized.iter().zip(&mut specialized_columns) {
                 if subpat.is_or_pat() {
-                    column.patterns.extend(subpat.iter_fields())
+                    column.patterns.extend(subpat.flatten_or_pat())
```

Comment:
It's kind of a shame that this code has an invariant (right?) that the `|` patterns must be flattened, but nothing protecting us against that in the first place.

Reply:
It is a shame indeed. I can see how I could encode it in the type system...

Reply:
I prefer not to actually. There are a lot of invariants in this file, this one is not one I'd rather spend my complexity budget on

Reply:
You got me thinking that I should really write down more of these invariants

Reply:
That would be awesome 😺